File: app/src/main/java/com/tasomaniac/openwith/util/IntentFixer.java
package com.tasomaniac.openwith.util;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ResolveInfo;
import android.net.Uri;

import java.util.List;

import static com.tasomaniac.openwith.extensions.AmazonASINKt.extractAmazonASIN;

public class IntentFixer {
    private static final Fixer[] INTENT_FIXERS = new Fixer[]{
            new AmazonFixer(),
    };

    public static Intent fixIntents(Context context, Intent intent) {

        for (Fixer intentFixer : INTENT_FIXERS) {
            Intent fixedIntent = intentFixer.fix(context, intent);

            if (hasHandler(context, fixedIntent)) {
                intent = fixedIntent;
            }
        }
        return intent;
    }

    /**
     * Queries on-device packages for a handler for the supplied {@link Intent}.
     */
    private static boolean hasHandler(Context context, Intent intent) {
        List<ResolveInfo> handlers = context.getPackageManager().queryIntentActivities(intent, 0);
        return !handlers.isEmpty();
    }

    interface Fixer {
        Intent fix(Context context, Intent intent);
    }

    private static class AmazonFixer implements Fixer {

        /**
         * If link contains amazon and it has ASIN in it, create a specific intent to open Amazon App.
         *
         * @param intent Original Intent with amazon link in it.
         * @return Specific Intent for Amazon app.
         */
        @Override
        public Intent fix(Context context, Intent intent) {
            if (intent.getDataString() != null && intent.getDataString().contains("amazon")) {
                String asin = extractAmazonASIN(intent.getDataString());
                if (asin != null) {
                    if ("0000000000".equals(asin)) {
                        return context.getPackageManager()
                                .getLaunchIntentForPackage(intent.getComponent().getPackageName());
                    }
                    return new Intent(Intent.ACTION_VIEW).setDataAndType(
                            Uri.parse("mshop://featured?ASIN=" + asin),
                            "vnd.android.cursor.item/vnd.amazon.mShop.featured"
                    );
                }
            }
            return intent;
        }
    }
}

File: resolver/src/main/kotlin/com/tasomaniac/openwith/util/IntentFixer.kt
package com.tasomaniac.openwith.util

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import com.tasomaniac.openwith.extensions.extractAmazonASIN

object IntentFixer {
    private val INTENT_FIXERS = arrayOf<Fixer>(
        AmazonFixer()
    )

    @JvmStatic
    fun fixIntents(context: Context, originalIntent: Intent): Intent {
        var intent = originalIntent
        for (intentFixer in INTENT_FIXERS) {
            val fixedIntent = intentFixer.fix(context, intent)
            if (context.packageManager.hasHandler(fixedIntent)) {
                intent = fixedIntent
            }
        }
        return intent
    }

    /**
     * Queries on-device packages for a handler for the supplied [intent].
     */
    private fun PackageManager.hasHandler(intent: Intent) = queryIntentActivities(intent, 0).isNotEmpty()

    internal interface Fixer {
        fun fix(context: Context, intent: Intent): Intent
    }

    private class AmazonFixer : Fixer {

        /**
         * If link contains amazon and it has ASIN in it, create a specific intent to open Amazon App.
         *
         * @param intent Original Intent with amazon link in it.
         * @return Specific Intent for Amazon app.
         */
        override fun fix(context: Context, intent: Intent): Intent {
            val dataString = intent.dataString
            if (dataString != null && dataString.contains("amazon")) {
                val asin = extractAmazonASIN(dataString)
                if (asin != null) {
                    return if ("0000000000" == asin) {
                        context.packageManager.getLaunchIntentForPackage(intent.component!!.packageName)!!
                    } else Intent(Intent.ACTION_VIEW).setDataAndType(
                        Uri.parse("mshop://featured?ASIN=$asin"),
                        "vnd.android.cursor.item/vnd.amazon.mShop.featured"
                    )
                }
            }
            return intent
        }
    }
}
